packages/big-design/src/components/ButtonGroup/ButtonGroup.tsx

```diff
@@ -0,0 +1,161 @@
+import { MoreHorizIcon } from '@bigcommerce/big-design-icons';
+import React, {
+  createRef,
+  HTMLAttributes,
+  isValidElement,
+  memo,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+
+import { useWindowResizeListener } from '../../hooks';
+import { MarginProps } from '../../mixins';
+import { ButtonProps } from '../Button';
+import { Dropdown } from '../Dropdown';
+import { Flex } from '../Flex';
+
+import { StyledButton, StyledFlexItem } from './styled';
+
+export interface ButtonGroupAction extends Omit<ButtonProps, 'children' | 'iconLeft' | 'iconRight'> {
+  text: string;
+}
+
+export interface ButtonGroupProps extends HTMLAttributes<HTMLDivElement>, MarginProps {
+  actions: ButtonGroupAction[];
+}
+
+export const ButtonGroup: React.FC<ButtonGroupProps> = memo(({ actions, ...wrapperProps }) => {
+  const parentRef = createRef<HTMLDivElement>();
+  const dropdownRef = createRef<HTMLDivElement>();
+  const [isMenuVisible, setIsMenuVisible] = useState(false);
+  const [actionsState, setActionsState] = useState(
+    actions.map((action) => ({
+      isVisible: true,
+      action,
+      ref: createRef<HTMLDivElement>(),
+    })),
+  );
+
+  const hideOverflowedElements = useCallback(() => {
+    const parentWidth = parentRef.current?.offsetWidth;
+    const dropdownWidth = dropdownRef.current?.offsetWidth;
+
+    if (!parentWidth || !dropdownWidth) {
+      return;
+    }
+
+    let remainingWidth = parentWidth;
+
+    const nextState = actionsState.map((stateObj) => {
+      const actionWidth = stateObj.ref.current?.offsetWidth;
+
+      if (!actionWidth) {
+        return stateObj;
+      }
+
+      if (remainingWidth - actionWidth > dropdownWidth) {
+        remainingWidth = remainingWidth - actionWidth;
+
+        return { ...stateObj, isVisible: true };
+      }
+
+      return { ...stateObj, isVisible: false };
+    });
+
+    const visibleActions = actionsState.filter(({ isVisible }) => isVisible);
+    const nextVisibleActions = nextState.filter(({ isVisible }) => isVisible);
+
+    if (visibleActions.length !== nextVisibleActions.length) {
+      setIsMenuVisible(nextVisibleActions.length !== actions.length);
+      setActionsState(nextState);
+    }
+  }, [actions, parentRef, dropdownRef, actionsState]);
+
+  const renderedDropdown = useMemo(() => {
+    const items = actionsState
+      .filter(({ isVisible }) => !isVisible)
+      .map((stateObj) => {
+        const action = stateObj.ref.current?.querySelector('button');
+        const icon = isValidElement(stateObj.action.iconOnly) ? stateObj.action.iconOnly : undefined;
+
+        return {
+          content: stateObj.action.text,
+          onItemClick: () => {
+            if (action) {
+              action.click();
+            }
+          },
+          hash: stateObj.action.text.toLowerCase(),
+          icon,
+        };
+      });
+
+    return (
+      <StyledFlexItem
+        data-testid="button-group-dropdown-toggle"
```

Comment:
As far as I know, we don't set `data-testid` in component directly, only when passed in as a prop during testing.

Reply:
I got PillTabs for the reference

Reply:
Chance and I had a very elaborate discussion around this, and in this case it is very tricky to write a good test without data-testids.. in this case its a hacky nightmare, basically because of the `offsetWidths`, and ended up keeping the data-testids... these are the only reliable way we could think of for targeting them and setting the offsetWidths so that we could test the collapsing functionality (jsdom does not actually render things, so all elements have an offsetWidth of 0 in the testing environment)